src/main/java/net/minecraftforge/event/PlayerConnectingEvent.java

```diff
@@ -0,0 +1,103 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.event;
+
+import com.mojang.authlib.GameProfile;
+import net.minecraft.network.NetworkManager;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.util.text.ITextComponent;
+import net.minecraft.util.text.TranslationTextComponent;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.eventbus.api.Cancelable;
+
+import javax.annotation.Nullable;
+
+/**
+ * ServerConnectEvent is fired when a player is connecting to a server right after the ban and whitelist checks. <br>
+ * <br>
+ * If the event is canceled, the player will be disconnected with either the set rejection message or a generic disconnected message.<br>
+ * <br>
+ * This event is {@link net.minecraftforge.eventbus.api.Cancelable}.<br>
+ * <br>
+ * This event does not have a result. {@link HasResult}<br>
+ * <br>
+ * This event is fired on the {@link MinecraftForge#EVENT_BUS}.
+ **/
+@Cancelable
+public class PlayerConnectingEvent extends net.minecraftforge.eventbus.api.Event {
+    private final MinecraftServer server;
+    private final NetworkManager networkManager;
+    private final GameProfile profile;
+    @Nullable
+    private final ITextComponent originalMessage;
+    @Nullable
+    private ITextComponent rejectionMessage;
+
+    public PlayerConnectingEvent(MinecraftServer server, NetworkManager networkManager, GameProfile profile, @Nullable ITextComponent original)
+    {
+        this.server = server;
+        this.networkManager = networkManager;
+        this.profile = profile;
+        this.originalMessage = original;
+        this.setRejectionMessage(original);
+    }
+
+    public NetworkManager getNetworkManager()
+    {
+        return networkManager;
+    }
+
+    public GameProfile getProfile()
+    {
+        return profile;
+    }
+
+    @Nullable
+    public ITextComponent getOriginalMessage()
+    {
+        return originalMessage;
+    }
+
+    @Nullable
+    public ITextComponent getRejectionMessage()
+    {
+        if (this.isCanceled() && this.rejectionMessage == null) {
+            return new TranslationTextComponent("multiplayer.disconnect.generic");
+        } else {
+            return rejectionMessage;
+        }
+    }
+
+    public void setRejectionMessage(@Nullable ITextComponent rejectionMessage)
+    {
+        if (rejectionMessage == null) {
+            this.setCanceled(false);
+            this.rejectionMessage = null;
+        } else {
+            this.setCanceled(true);
+            this.rejectionMessage = rejectionMessage;
+        }
+    }
```

Comment:
One thing that was [noted in a previous PR](https://github.com/MinecraftForge/MinecraftForge/pull/6158#issuecomment-531923415) is that people should probably not be able to turn an already rejected login into an accepted one.
I agree with that. I might even go so far as to say to not even fire the event if vanilla rejects the message: Do we really want mods to be able to bypass the vanilla bans and whitelist? Probably not.

Reply:
Being able to modify the vanilla whitelist and bans is the main point why I made this PR, say you have a mod that adds additional whitelist.jsons from various sources, if the player isn't in the vanilla whitelist the mod wouldn't be able to do that.

Reply:
The vanilla whitelist should just be disabled in that case.

Reply:
Idealy the event would let you know if the vanilla action was to block and let you overwrite that.

> The vanilla whitelist should just be disabled in that case.

I get many, many support requests for my mod (MCLink) from people who enable the whitelist and expect my mod to then overwrite that setting. I could overwrite that option on every startup, overwrite the command, and do a whole bunch of other, non-user friendly things. But I'd prefer not to do that and just have the event always fire and let the modder handle it.

Reply:
I don't like the idea of mods being able to bypass vanilla whitelist decisions.
If as a user I enable the whitelist I expect it to be enforced. If I don't want it enabled, I don't enable it.

Reply:
You are not an average user though. An average user expects to install a "whitelist mod" on their server and everything to Just Work™. Explaining to them that they have to install a "whitelisting mod" and then _disable_ the whitelist results in confusion.

Then you also get the added bonus of the following scenario (has happend, I've received complaints about it):
- User has a whitelisted server
- User installs whitelist mod to do some automatic whitelist managent
- User has to disalbe the whitelist because of reasons
- Everything works for a wile
- User decides to remove the mod for any reason
- Server is now no longer protected by any whitelist because they had to disable it previously

I agree that this has potential to be abused. But so do many, many other things. And when they do get abused, people stop using the mod or they complain.
I'd say make it a "paranoid whitelist" config option, but honestly, that's gonna be even more confusing.

Reply:
Having a mod-specific "lockout whitelist" is confusing to users by itself.
I'd expect to be able to enable the vanilla whitelist when maintenance on a server is required and not have to search through (potentially multiple) mod config files.
A mod might not even implement a method to have a "lockout" resulting in a user having to enable/disable mods to achieve a basic task which would be otherwise accomplished by simply typing `whitelist on`.